### ISSUE TITLE
TypeScript doesn't narrow out `undefined` after constructor call when callee is `any`

### ISSUE THREAD

**Issue by fatcerberus** [Microsoft/TypeScript#27993]
TypeScript doesn't seem to narrow the type of a variable correctly in cases where the constructor being called is typed as `any`.  If the type of a variable includes `undefined` in its type signature, it's not narrowed out after a constructor call even though by definition (I think--correct me if I'm wrong) `new fn()` *must* always evaluate to an object if it doesn't throw first.

**TypeScript Version:**  3.1.3

**Search Terms:** constructor narrow

**Code**

```ts
let x: { [x: string]: any } | undefined;
x = new (Object as any)();
// AFAIK `x` can't possibly be undefined after this point
console.log(x.foo);  // error: Object is possibly 'undefined'.
```

**Expected behavior:**
No error.

**Actual behavior:**
Compile-time error: `Object is possibly 'undefined'.`

**Playground Link:**
http://www.typescriptlang.org/play/#src=let%20x%3A%20%7B%20%5Bx%3A%20string%5D%3A%20any%20%7D%20%7C%20undefined%3B%0D%0Ax%20%3D%20new%20(Object%20as%20any)()%3B%0D%0Aconsole.log(x.foo)%3B%0D%0A

**Comment by weswigham** [comment]
Technically this is because `any` includes `undefined`. So when we narrow `{ [x: string]: any } | undefined;` to only members assignable to `any`, we get.... `{ [x: string]: any } | undefined;`. It's a decent argument for using `subtype` and not `assignability` in `getAssignmentReducedType`.

**Comment by weswigham** [comment]
Rather than casting like `new (Object as any)()` you could cast as `new (Object as new() => object)()`? Or `new (Object as new() => {[i: string]: any})()`. I know that technically a constructor can't return `undefined` (since an `undefined` return just means to use the supplied `this` as the result), so technically constructing an `any` doesn't yield `any` but rather `any - undefined`, but we have pretty much no way to represent that in the typesystem.

**Comment by fatcerberus** [comment]
To be clear: The `new (Object as any)()` was just to illustrate the issue in the standalone repro.  In my actual use case the constructor in question comes from a dynamic import where it is already typed as `any` and I was trying to avoid a cast (since I’d already done checks to verify it’s constructible beforehand).  It just surprised me that the compiler didn’t realize `new Anything()` can never be `undefined` is all.

If a cast is the only workaround, so be it. :smiley:

**Comment by fatcerberus** [comment]
Instead of a cast, would a non-null assertion be enough to squelch the error?

```ts
x = new (Object as any)()!;
x.foo;
```

**Comment by weswigham** [comment]
I don't think so, no; for the same reason I said above about subtraction types - `NonNullable<any>` is still `any` because we have no way of representing the removal of a type from a broader type's domain.

**Comment by fatcerberus** [comment]
I guess `any` is kind of like the type system’s version of infinity - infinity minus anything is still infinity.  That makes sense now, thanks.

But now I finally understand the underlying issue: `new <any>()` is being inferred as `any` which flies in the face of expectation since I know that per specification, calling any random value in a program using `new` can only have two possible results: it will either throw (and therefore the type doesn’t matter), or else return something which is assignable to:
```ts
interface ICtorResult {
    [x: string]: any
    [x: number]: any
    [x: symbol]: any
}
```

...and therefore any `new` expression, regardless of what’s being new’d, can safely be inferred as some intersection of its known type and the interface above.  While *subtracting* from infinity doesn’t work, the *intersection* of “everything” with “only some things”, mathetically speaking, is “only some things”.  Could the type system be made to do *that*?

**Comment by ethanresnick** [comment]
Can negated types be used to solve this?

**Comment by deser** [comment]
> I don't think so, no; for the same reason I said above about subtraction types - `NonNullable<any>` is still `any` because we have no way of representing the removal of a type from a broader type's domain.

Why? It seems quite reasonable from a typescript user point of view to say: hey, I want that function returns anything despite undefined | null

**Comment by fatcerberus** [comment]
@deser Nobody said it wasn't *useful*; it's just not *possible* today.  Somebody has to implement it first.
